### DISCUSSION TITLE
Ref/Out overrides don't follow spec

### DISCUSSION THREAD

**Discussion by yaakov-h** [dotnet/csharplang#1255] [General]
My understanding of the C# spec is that `ref` and `out` are treated separately for method overload resolution:

> For other purposes of signature matching (e.g., hiding or overriding), `ref` and `out` are considered part of the signature and do not match each other. - [Signatures and overloading](https://github.com/dotnet/csharplang/blob/master/spec/basic-concepts.md#signatures-and-overloading)

Given the following code, `Baz.M(ref object)` should override `Foo.M(ref object)`, but at runtime it appears to override `Bar(out object)` instead.

"Go To Definition" agrees with me, or has made the same misunderstanding (see the comment below).

```cs
using System;

public static class Program
{
    public static void Main()
    {
        object o = null;
        new Baz().M(ref o); // Should print "Baz"
        new Baz().M(out o); // Should print "Bar"
        new Baz().N(0); // Should print "Baz"
    }
}

class Foo
{
    public virtual void M(ref object o)
    {
        Console.WriteLine("Foo");
    }

    public virtual void N(int i)
    {
        Console.WriteLine("Foo");
    }
}

class Bar : Foo
{
    // New method, does not override Foo.M()        
    public virtual void M(out object o)
    {
        o = null;
        Console.WriteLine("Bar");
    }

    // New method, does not override Foo.N()
    public virtual void N(int i)
    {
        Console.WriteLine("Bar");
    }
}

class Baz : Bar
{
    // Should override Foo.M()
    public override void M(ref object o)
    {
        // If you uncomment this line, then Go To Definition goes to Foo.M()
        // base.M(ref o);

        Console.WriteLine("Baz");
    }

    // Should override Bar.N()
    public override void N(int i)
    {
        Console.WriteLine("Baz");
    }
}
```

**Expected Output:**

```
Baz
Bar
Baz
```

**Actual Output:**

```
Foo
Baz
Baz
```

Is this:

* A complete misunderstanding on my part?
* A bug in Go To Definition?
* An error in the language spec?
* A bug in Roslyn?
* A bug in the .NET Framework and .NET Core runtimes?

Thanks.

**Comment by Thaina** [comment]
Interesting. I also play with it a little more and found out that

![image](https://user-images.githubusercontent.com/1042507/34930995-df8328a6-f9fe-11e7-8672-31986bea44b2.png)

Might be related

**Comment by yaakov-h** [comment]
Yep, the spec mentions that too, right before the bit above:

> Although `out` and `ref` parameter modifiers are considered part of a signature, members declared in a single type cannot differ in signature solely by `ref` and `out`. A compile-time error occurs if two members are declared in the same type with signatures that would be the same if all parameters in both methods with `out` modifiers were changed to `ref` modifiers.

**Comment by Thaina** [comment]
@yaakov-h If so then I don't think this is a bug. ` public virtual void M(out object o)` in `Bar` just has the effect as putting `new` keyword to replace the original `Foo.M` because it has exactly same signature

Maybe a bug in compiler warning that it not warn anything about this

**Comment by yaakov-h** [comment]
Maybe, except the spec explicitly says that for the purposes of hiding or overriding - which is exactly what we are doing here - "`ref` and `out` are considered part of the signature and do not match each other."

**Comment by Thaina** [comment]
@yaakov-h I see then I think it was a bug from the point that `Bar.M` was not error from being hiding `Foo.M`

**Comment by MkazemAkhgary** [comment]
have you tried this on older c# compilers? I'm wondering for how long it was unnoticed.

**Comment by yaakov-h** [comment]
According to a coworker this behaviour has been around for years.

**Comment by stakx** [comment]
> have you tried this on older c# compilers? I'm wondering for how long it was unnoticed.

I just tried this with Mono's C# compiler, and with http://rextester.com which claims to use C# compiler version 4.0.30319.17929. (If I understand correctly, this is a pre-Roslyn version of the C# compiler.) They all produce the same output as shown above. So this behaviour has probably been around for a long time.

It appears to me that when it comes to "signatures", the C# spec differs from the underlying platform's notion of signatures. As far as I can tell, CLI signatures (at least the standalone ones) do not distinguish between `ref` and `out`&mdash;there are both simply by-ref parameters. (See ECMA-335, §22.2.10.) The CLI considers optional or required modifiers as being part of a signature, but C# doesn't use these to mark `ref` or `out` parameters. And custom attributes aren't considered part of a signature, so distinguishing using the `[Out]` pseudo-attribute is out of the question, too.

The C# spec OTOH explicitly states (as mentioned in the initial post) that `ref` and `out` don't match each other with regard to overriding / method hiding.

Perhaps the C# compiler simply takes a shortcut here and implements its overloading / method hiding rules by directly employing the CLI's `hidebysig` mechanism unaltered (which might be good enough in practice, but not really sufficient given the differences in the respective specifications)? The C# spec suggests just that in §3.6 Signatures and overloading:

> This restriction *[not being able to overload based on `ref` or `out` alone in the context of a single type]* is to allow C# programs to be easily translated to run on the Common Language Infrastructure (CLI), which does not provide a way to define methods that differ solely in `ref` and `out`.
> ...
> `F(ref int)` and `F(out int)` cannot be declared within the same interface because their signatures differ solely by `ref` and `out`.

Could it be that the spec writers simply forgot to consider cases where such signature "collisions" appear, not in a single type, but in a type hierarchy?

**Comment by brian-reichle** [comment]
>   have you tried this on older c# compilers? I'm wondering for how long it was unnoticed.

I first noticed it around the .NET 4.0/4.5 days, I'm fairly sure I also tested it on a machine that only had .NET 3.5 and VisualStudio 2008 installed (I don't recall if csc was distributed with the framework or visual studio back then). My interpretation of ECMA-334 at the time was that having the methods in the same type was prohibited but different types in the same inheritance chain was ok (if you don't mind crazy code).

As @stakx pointed out, the ref and out are distinguished by an attribute and attributes are not considered part of the signature by the runtime, though I have seen various tools behave as though it were.

I believe this could be resolved with a MethodImpl record, but I think that might be heading down a problematic path and I would be happy if the C# spec were simply 'clarified' to prohibit this :)

**Comment by stakx** [comment]
> the ref and out are distinguished by an attribute

(@brian-reichle - I need to correct myself there: AFAIK, it isn't actually `OutAttribute` itself. That is only a pseudo-custom attribute and doesn't even get emitted as such; it is turned into a particular flag (`ParamAttributes.Out`) in the Params metadata table, see ECMA-335 §20.2.1 and §22.1.12. The information is there in IL metadata, but the runtime probably doesn't use it. I cannot find the exact definition of the `hidebysig` signature matching rules right now, but I suspect it doesn't take any `ParamAttributes` into account, so the end result is the same.)

**Comment by brian-reichle** [comment]
@stakx, I think you might be right, but either way it's not using a modopt or modreq so the end result is the same. IIRC, matching (for the purpose of overriding) is based only on the method name, and signature blob, ParamAttributes.Out is definitely not part of the signature blob.

hidebysig is ignored by the runtime (Partition II section 15.4.2), I believe it's supposed to provide guidance for tools so they can match language behaviour.

**Comment by brian-reichle** [comment]
Just for the record, here is a/the error report on connect from 2014

https://connect.microsoft.com/VisualStudio/feedback/details/815996/c-compiler-out-vs-ref-override-added-to-wrong-vtable-slot

**Comment by tannergooding** [comment]
FYI. @vsadov, @jcouv

**Comment by jcouv** [comment]
From the [readonly-ref](https://github.com/dotnet/csharplang/blob/master/proposals/csharp-7.2/readonly-ref.md) spec:
> It is not permitted to overload on ref/out/in differences.
> It is permitted to overload on ordinary byval and in differences.

Moving to roslyn repo as a compiler bug.

**Comment by jcouv** [comment]
Issue moved to [dotnet/roslyn #24347](https://github.com/dotnet/roslyn/issues/24347) via [**ZenHub**](https://www.zenhub.com/)

**Comment by jcouv** [comment]
Re-opened, as this turns out not be a compiler bug. Copying comment from @VSadov 

---------
This is a long standing issue with a design choice that virtual calls in C# are emitted via "least derived" instead of the "most derived" method.

It is confusing in cases where inheritance hierarchy as seen by CLR is not the same as seen by C#. 
Here CLR does not understand ref/out differences, but C# does. 

So when user sees call to `Baz.M`, which is intuitive and is supported by the VS and GTD, in reality a call to `Foo.M` is emitted. 
Essentially C# traces the target method back to its least derived override, hoping that CLR will trace it back to the most derived one in the context of the call. However, since C# and CLR disagree on the hierarchy, CLR arrives to to a different method and calls that.

With introduction of named parameters, optional parameters, tuples and other features where the information from the most derived statically known method wins over, such behavior is getting more and more confusing.

Every time we see a bug like this we consider a change where we emit calls via the _most_ derived override. In most cases the difference would be unobservable, but the confusing cases like this would be "fixed" and that in itself could be considered a breaking change.

I think the evidence is mounting that we should make the fix. 

This should go back to the dotnet/csharplang.

CC:@gafter

**Comment by 4creators** [comment]
> This is a long standing issue with a design choice that virtual calls in C# are emitted via "least derived" instead of the "most derived" method.

Seems that fix for that C#/Roslyn/CLR joint misbehavior is overdue as the issue is not that rare. It would support significantly improved overload resolution in future and increase language and runtime flexibility in implementing other features which could impact overload resolution: see  #886.

**Comment by VSadov** [comment]
@jcouv - probably need to change the title. The binder and overload/override resolution are correct here and respect the spec.

The IL that we emit, however, could be more robust. Emitting calls against the most derived target would reduce the chances that something unexpected gets called.

**Comment by VSadov** [comment]
The fix would not do anything to the binding. It is basically about removing the following line of code:

https://github.com/dotnet/roslyn/blob/3bf3a102e61c1e12ad7d6b9f05c1bc6d4eab523a/src/Compilers/CSharp/Portable/CodeGen/EmitExpression.cs#L1557

**Comment by brian-reichle** [comment]
I don't believe that switching to the most derived method would fix this specific issue. It might make the call to `Baz.M(ref object)` appear to be correct, but the implementation of `Baz.M(ref object)` will still be placed in the virtual method table slot for `Bar.M(out object)` and so the call to `Bar.M(out object)` will still be incorrect.
